xayn-ai/src/coi/system.rs

```diff
@@ -214,6 +223,160 @@ impl CoiSystem {
+            coi.key_phrases()
+                .iter()
+                .chain(candidates.iter())
+                .map(|key_phrase| key_phrase.point().view())
+                .chain(once(coi.point().view())),
+        )
+        .slice_move(s![..len, ..]);
+        debug_assert!(similarity.iter().copied().all(f32::is_finite));
+
+        let min = reduce_without_diag(similarity.view(), Axis(0), f32::min, identity);
+        let max = reduce_without_diag(similarity.view(), Axis(0), f32::max, identity);
+        let normalized = (&similarity - &min) / (max - min);
+        let mean = reduce_without_diag(
+            normalized.view(),
+            Axis(0),
+            |reduced, element| reduced + element,
+            |reduced| reduced / (len - 1) as f32,
+        );
+        let std_dev = reduce_without_diag(
+            &normalized - &mean,
+            Axis(0),
+            |reduced, element| reduced + element.powi(2),
+            |reduced| (reduced / (len - 1) as f32).sqrt(),
+        );
+        let normalized = (normalized - mean) / std_dev + 0.5;
+        let normalized = normalized
+            .mapv_into(|normalized| normalized.is_finite().then(|| normalized).unwrap_or(0.5));
+        debug_assert!(normalized.iter().copied().all(f32::is_finite));
+
+        let candidate = argmax(normalized.slice(s![.., len]));
+        let mut selected = vec![false; len];
+        selected[candidate] = true;
+        for _ in 0..max_key_phrases - 1 {
+            let candidate = argmax(selected.iter().zip(normalized.rows()).map(
+                |(&is_selected, normalized)| {
+                    if is_selected {
+                        f32::MIN
+                    } else {
+                        let max = selected
+                            .iter()
+                            .zip(normalized)
+                            .filter_map(|(is_selected, normalized)| {
+                                is_selected.then(|| *normalized)
+                            })
+                            .reduce(f32::max)
+                            .unwrap(/* at least one key phrase is selected */);
+                        self.config.gamma * normalized[len] - (1. - self.config.gamma) * max
+                    }
+                },
+            ));
+            selected[candidate] = true;
+        }
+
+        let relevance = selected
+            .iter()
+            .zip(similarity.slice(s![.., len]))
+            .filter_map(|(is_selected, similarity)| is_selected.then(|| similarity))
+            .copied();
+        let max =
+            relevance.clone().reduce(f32::max).unwrap(/* at least one key phrase is selected */);
+        let relevance = relevance.map(|relevance| {
+            (relevance > 0.)
+                .then(|| (relevance / max).clamp(0., 1.))
+                .unwrap_or_default()
+        });
+
+        let key_phrases = coi.swap_key_phrases(BTreeSet::default());
+        let key_phrases = selected
```

Comment:
Also here we can avoid this zip chain, if the index in `selected` is less then `key_pharases.len()` we need to take it from here otherwise from `candidates`, then we can access the relevance by index and set the new keyphrases instead of swap.

Reply:
sounds good, i'll give it a try.

Reply:
as discussed, i'll tackle this in another pr.

Reply:
#361